File: mobile-sdk/delivez_api_client.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

/// Delivez Mobile API Client (Flutter / Dart)
/// Ready-to-use production client for mobile apps.
class DelivezApiClient {
  final String baseUrl;
  String? _accessToken;

  DelivezApiClient({
    this.baseUrl = 'http://localhost:4000/api/v1',
    String? initialToken,
  }) : _accessToken = initialToken;

  void setAccessToken(String? token) {
    _accessToken = token;
  }

  Map<String, String> _buildHeaders([Map<String, String>? extra]) {
    final headers = {
      'Accept': 'application/json',
      'Content-Type': 'application/json',
      'X-Device-Platform': 'Flutter',
    };
    if (_accessToken != null && _accessToken!.isNotEmpty) {
      headers['Authorization'] = 'Bearer $_accessToken';
    }
    if (extra != null) {
      headers.addAll(extra);
    }
    return headers;
  }

  dynamic _handleResponse(http.Response response) {
    final body = jsonDecode(response.body);
    if (response.statusCode >= 200 && response.statusCode < 300) {
      return body['data'];
    } else {
      final msg = body['message'] ?? 'Request failed with status ${response.statusCode}';
      throw DelivezApiException(response.statusCode, msg, body['error']);
    }
  }

  // ==========================================
  // 1. SYSTEM & HOME
  // ==========================================
  Future<Map<String, dynamic>> checkHealth() async {
    final res = await http.get(Uri.parse('$baseUrl/health'), headers: _buildHeaders());
    return _handleResponse(res);
  }

  Future<Map<String, dynamic>> getHomeFeed() async {
    final res = await http.get(Uri.parse('$baseUrl/home/all'), headers: _buildHeaders());
    return _handleResponse(res);
  }

  Future<Map<String, dynamic>> detectLocation([double? lat, double? lng]) async {
    String endpoint = '$baseUrl/location/current';
    if (lat != null && lng != null) {
      endpoint = '$baseUrl/location/detect?lat=$lat&lng=$lng';
    }
    final res = await http.get(Uri.parse(endpoint), headers: _buildHeaders());
    return _handleResponse(res);
  }

  Future<Map<String, dynamic>> checkPincode(String pincode) async {
    final res = await http.get(Uri.parse('$baseUrl/pincode/$pincode'), headers: _buildHeaders());
    return _handleResponse(res);
  }

  // ==========================================
  // 2. AUTHENTICATION & PROFILE
  // ==========================================
  Future<Map<String, dynamic>> sendLoginOtp(String mobileNumber, {String countryCode = '+91'}) async {
    final res = await http.post(
      Uri.parse('$baseUrl/auth/login'),
      headers: _buildHeaders(),
      body: jsonEncode({'countryCode': countryCode, 'mobileNumber': mobileNumber}),
    );
    return _handleResponse(res);
  }

  Future<Map<String, dynamic>> verifyOtp(String challengeId, String otp, {String? deviceId}) async {
    final res = await http.post(
      Uri.parse('$baseUrl/auth/verify-otp'),
      headers: _buildHeaders(),
      body: jsonEncode({'challengeId': challengeId, 'otp': otp, 'deviceId': deviceId}),
    );
    final data = _handleResponse(res);
    if (data['accessToken'] != null) {
      setAccessToken(data['accessToken']);
    }
    return data;
  }

  Future<Map<String, dynamic>> getProfile() async {
    final res = await http.get(Uri.parse('$baseUrl/auth/me'), headers: _buildHeaders());
    return _handleResponse(res);
  }

  // ==========================================
  // 3. SAVED ADDRESSES
  // ==========================================
  Future<List<dynamic>> getAddresses() async {
    final res = await http.get(Uri.parse('$baseUrl/addresses'), headers: _buildHeaders());
    final data = _handleResponse(res);
    return data['addresses'] ?? [];
  }

  Future<Map<String, dynamic>> saveAddress(Map<String, dynamic> addressPayload) async {
    final res = await http.post(
      Uri.parse('$baseUrl/addresses'),
      headers: _buildHeaders(),
      body: jsonEncode(addressPayload),
    );
    return _handleResponse(res);
  }

  // ==========================================
  // 4. SERVICE 1: PERSONAL COURIER
  // ==========================================
  Future<Map<String, dynamic>> getCourierOptions() async {
    final res = await http.get(Uri.parse('$baseUrl/courier/options'), headers: _buildHeaders());
    return _handleResponse(res);
  }

  Future<Map<String, dynamic>> getCourierQuote(Map<String, dynamic> quotePayload) async {
    final res = await http.post(
      Uri.parse('$baseUrl/courier/quote'),
      headers: _buildHeaders(),
      body: jsonEncode(quotePayload),
    );
    return _handleResponse(res);
  }

  Future<Map<String, dynamic>> createCourierBooking(Map<String, dynamic> bookingPayload) async {
    final res = await http.post(
      Uri.parse('$baseUrl/courier/bookings'),
      headers: _buildHeaders({'Idempotency-Key': 'idemp-${DateTime.now().millisecondsSinceEpoch}'}),
      body: jsonEncode(bookingPayload),
    );
    return _handleResponse(res);
  }

  Future<Map<String, dynamic>> trackCourierBooking(String id) async {
    final res = await http.get(Uri.parse('$baseUrl/courier/bookings/$id/track'), headers: _buildHeaders());
    return _handleResponse(res);
  }

  Future<Map<String, dynamic>> getCourierPod(String id) async {
    final res = await http.get(Uri.parse('$baseUrl/courier/bookings/$id/pod'), headers: _buildHeaders());
    return _handleResponse(res);
  }

  // ==========================================
  // 5. SERVICE 2: LUGGAGE DELIVERY
  // ==========================================
  Future<Map<String, dynamic>> getLuggageOptions() async {
    final res = await http.get(Uri.parse('$baseUrl/luggage-delivery/options'), headers: _buildHeaders());
    return _handleResponse(res);
  }

  Future<Map<String, dynamic>> getLuggageQuote(Map<String, dynamic> quotePayload) async {
    final res = await http.post(
      Uri.parse('$baseUrl/luggage-delivery/quote'),
      headers: _buildHeaders(),
      body: jsonEncode(quotePayload),
    );
    return _handleResponse(res);
  }

  Future<Map<String, dynamic>> createLuggageBooking(Map<String, dynamic> bookingPayload) async {
    final res = await http.post(
      Uri.parse('$baseUrl/luggage-delivery/bookings'),
      headers: _buildHeaders({'Idempotency-Key': 'idemp-${DateTime.now().millisecondsSinceEpoch}'}),
      body: jsonEncode(bookingPayload),
    );
    return _handleResponse(res);
  }

  // ==========================================
  // 6. SERVICE 3: CONFIDENTIAL VAULT DELIVERY
  // ==========================================
  Future<Map<String, dynamic>> getVaultOptions() async {
    final res = await http.get(Uri.parse('$baseUrl/confidential-delivery/options'), headers: _buildHeaders());
    return _handleResponse(res);
  }

  Future<Map<String, dynamic>> getVaultQuote(Map<String, dynamic> quotePayload) async {
    final res = await http.post(
      Uri.parse('$baseUrl/confidential-delivery/quote'),
      headers: _buildHeaders(),
      body: jsonEncode(quotePayload),
    );
    return _handleResponse(res);
  }

  Future<Map<String, dynamic>> createVaultBooking(Map<String, dynamic> bookingPayload) async {
    final res = await http.post(
      Uri.parse('$baseUrl/confidential-delivery/bookings'),
      headers: _buildHeaders({'Idempotency-Key': 'idemp-${DateTime.now().millisecondsSinceEpoch}'}),
      body: jsonEncode(bookingPayload),
    );
    return _handleResponse(res);
  }

  Future<Map<String, dynamic>> trackVaultShipment(String vaultId) async {
    final res = await http.get(Uri.parse('$baseUrl/confidential-delivery/track/$vaultId'), headers: _buildHeaders());
    return _handleResponse(res);
  }

  Future<Map<String, dynamic>> verifyVaultOtp(String id, String otp) async {
    final res = await http.post(
      Uri.parse('$baseUrl/confidential-delivery/track/$id/verify-otp'),
      headers: _buildHeaders(),
      body: jsonEncode({'otp': otp}),
    );
    return _handleResponse(res);
  }

  // ==========================================
  // 7. SERVICE 4: FORGOT SOMETHING
  // ==========================================
  Future<Map<String, dynamic>> getForgotSomethingOptions() async {
    final res = await http.get(Uri.parse('$baseUrl/forgot-something/options'), headers: _buildHeaders());
    return _handleResponse(res);
  }

  Future<Map<String, dynamic>> getForgotSomethingQuote(Map<String, dynamic> quotePayload) async {
    final res = await http.post(
      Uri.parse('$baseUrl/forgot-something/quote'),
      headers: _buildHeaders(),
      body: jsonEncode(quotePayload),
    );
    return _handleResponse(res);
  }

  Future<Map<String, dynamic>> createForgotSomethingBooking(Map<String, dynamic> bookingPayload) async {
    final res = await http.post(
      Uri.parse('$baseUrl/forgot-something/bookings'),
      headers: _buildHeaders({'Idempotency-Key': 'idemp-${DateTime.now().millisecondsSinceEpoch}'}),
      body: jsonEncode(bookingPayload),
    );
    return _handleResponse(res);
  }

  Future<Map<String, dynamic>> trackForgotSomething(String id) async {
    final res = await http.get(Uri.parse('$baseUrl/forgot-something/track/$id'), headers: _buildHeaders());
    return _handleResponse(res);
  }

  Future<Map<String, dynamic>> verifyForgotSomethingOtp(String id, String type, String otp) async {
    final res = await http.post(
      Uri.parse('$baseUrl/forgot-something/track/$id/verify-otp'),
      headers: _buildHeaders(),
      body: jsonEncode({'type': type, 'otp': otp}),
    );
    return _handleResponse(res);
  }

  // ==========================================
  // 8. UNIVERSAL TRACKER & SUPPORT
  // ==========================================
  Future<Map<String, dynamic>> universalTrack(String trackingId) async {
    final res = await http.get(Uri.parse('$baseUrl/track/$trackingId'), headers: _buildHeaders());
    return _handleResponse(res);
  }

  Future<Map<String, dynamic>> getSupportConfig() async {
    final res = await http.get(Uri.parse('$baseUrl/support/config'), headers: _buildHeaders());
    return _handleResponse(res);
  }

  Future<Map<String, dynamic>> submitInquiry(Map<String, dynamic> inquiryPayload) async {
    final res = await http.post(
      Uri.parse('$baseUrl/support/inquiry'),
      headers: _buildHeaders(),
      body: jsonEncode(inquiryPayload),
    );
    return _handleResponse(res);
  }
}

class DelivezApiException implements Exception {
  final int statusCode;
  final String message;
  final dynamic details;

  DelivezApiException(this.statusCode, this.message, [this.details]);

  @override
  String toString() => 'DelivezApiException($statusCode): $message';
}

File: mobile-sdk/delivezApiClient.ts
/**
 * Delivez Mobile API Client (React Native / TypeScript)
 * Ready-to-use production client for React Native mobile apps.
 */
export class DelivezApiClient {
  private baseUrl: string;
  private accessToken: string | null = null;

  constructor(baseUrl: string = 'http://localhost:4000/api/v1', initialToken?: string) {
    this.baseUrl = baseUrl;
    if (initialToken) this.accessToken = initialToken;
  }

  public setAccessToken(token: string | null) {
    this.accessToken = token;
  }

  private async request<T = any>(
    path: string,
    options: RequestInit = {}
  ): Promise<T> {
    const headers: Record<string, string> = {
      Accept: 'application/json',
      'Content-Type': 'application/json',
      'X-Device-Platform': 'ReactNative',
      ...((options.headers as Record<string, string>) || {}),
    };

    if (this.accessToken) {
      headers['Authorization'] = `Bearer ${this.accessToken}`;
    }

    const url = `${this.baseUrl}/${path.replace(/^\//, '')}`;
    const response = await fetch(url, { ...options, headers });
    const json = await response.json();

    if (!response.ok) {
      throw new DelivezApiError(
        response.status,
        json.message || 'API request failed',
        json.error
      );
    }

    return json.data as T;
  }

  // ==========================================
  // 1. SYSTEM & HOME
  // ==========================================
  public checkHealth() {
    return this.request('/health');
  }

  public getHomeFeed() {
    return this.request('/home/all');
  }

  public detectLocation(lat?: number, lng?: number) {
    if (lat !== undefined && lng !== undefined) {
      return this.request(`/location/detect?lat=${lat}&lng=${lng}`);
    }
    return this.request('/location/current');
  }

  public checkPincode(pincode: string) {
    return this.request(`/pincode/${pincode}`);
  }

  // ==========================================
  // 2. AUTHENTICATION & PROFILE
  // ==========================================
  public sendLoginOtp(mobileNumber: string, countryCode: string = '+91') {
    return this.request('/auth/login', {
      method: 'POST',
      body: JSON.stringify({ countryCode, mobileNumber }),
    });
  }

  public async verifyOtp(challengeId: string, otp: string, deviceId?: string) {
    const data = await this.request<{ user: any; accessToken: string }>('/auth/verify-otp', {
      method: 'POST',
      body: JSON.stringify({ challengeId, otp, deviceId }),
    });
    if (data?.accessToken) {
      this.setAccessToken(data.accessToken);
    }
    return data;
  }

  public getProfile() {
    return this.request('/auth/me');
  }

  // ==========================================
  // 3. SAVED ADDRESSES
  // ==========================================
  public async getAddresses() {
    const data = await this.request<{ addresses: any[] }>('/addresses');
    return data.addresses || [];
  }

  public saveAddress(address: Record<string, any>) {
    return this.request('/addresses', {
      method: 'POST',
      body: JSON.stringify(address),
    });
  }

  // ==========================================
  // 4. SERVICE 1: PERSONAL COURIER
  // ==========================================
  public getCourierOptions() {
    return this.request('/courier/options');
  }

  public getCourierQuote(payload: Record<string, any>) {
    return this.request('/courier/quote', {
      method: 'POST',
      body: JSON.stringify(payload),
    });
  }

  public createCourierBooking(payload: Record<string, any>) {
    return this.request('/courier/bookings', {
      method: 'POST',
      headers: { 'Idempotency-Key': `idemp-${Date.now()}` },
      body: JSON.stringify(payload),
    });
  }

  public trackCourierBooking(id: string) {
    return this.request(`/courier/bookings/${id}/track`);
  }

  public getCourierPod(id: string) {
    return this.request(`/courier/bookings/${id}/pod`);
  }

  // ==========================================
  // 5. SERVICE 2: LUGGAGE DELIVERY
  // ==========================================
  public getLuggageOptions() {
    return this.request('/luggage-delivery/options');
  }

  public getLuggageQuote(payload: Record<string, any>) {
    return this.request('/luggage-delivery/quote', {
      method: 'POST',
      body: JSON.stringify(payload),
    });
  }

  public createLuggageBooking(payload: Record<string, any>) {
    return this.request('/luggage-delivery/bookings', {
      method: 'POST',
      headers: { 'Idempotency-Key': `idemp-${Date.now()}` },
      body: JSON.stringify(payload),
    });
  }

  // ==========================================
  // 6. SERVICE 3: CONFIDENTIAL VAULT DELIVERY
  // ==========================================
  public getVaultOptions() {
    return this.request('/confidential-delivery/options');
  }

  public getVaultQuote(payload: Record<string, any>) {
    return this.request('/confidential-delivery/quote', {
      method: 'POST',
      body: JSON.stringify(payload),
    });
  }

  public createVaultBooking(payload: Record<string, any>) {
    return this.request('/confidential-delivery/bookings', {
      method: 'POST',
      headers: { 'Idempotency-Key': `idemp-${Date.now()}` },
      body: JSON.stringify(payload),
    });
  }

  public trackVaultShipment(vaultId: string) {
    return this.request(`/confidential-delivery/track/${vaultId}`);
  }

  public verifyVaultOtp(id: string, otp: string) {
    return this.request(`/confidential-delivery/track/${id}/verify-otp`, {
      method: 'POST',
      body: JSON.stringify({ otp }),
    });
  }

  // ==========================================
  // 7. SERVICE 4: FORGOT SOMETHING
  // ==========================================
  public getForgotSomethingOptions() {
    return this.request('/forgot-something/options');
  }

  public getForgotSomethingQuote(payload: Record<string, any>) {
    return this.request('/forgot-something/quote', {
      method: 'POST',
      body: JSON.stringify(payload),
    });
  }

  public createForgotSomethingBooking(payload: Record<string, any>) {
    return this.request('/forgot-something/bookings', {
      method: 'POST',
      headers: { 'Idempotency-Key': `idemp-${Date.now()}` },
      body: JSON.stringify(payload),
    });
  }

  public trackForgotSomething(id: string) {
    return this.request(`/forgot-something/track/${id}`);
  }

  public verifyForgotSomethingOtp(id: string, type: 'PICKUP' | 'DELIVERY', otp: string) {
    return this.request(`/forgot-something/track/${id}/verify-otp`, {
      method: 'POST',
      body: JSON.stringify({ type, otp }),
    });
  }

  // ==========================================
  // 8. UNIVERSAL TRACKER & SUPPORT
  // ==========================================
  public universalTrack(trackingId: string) {
    return this.request(`/track/${encodeURIComponent(trackingId)}`);
  }

  public getSupportConfig() {
    return this.request('/support/config');
  }

  public submitInquiry(payload: Record<string, any>) {
    return this.request('/support/inquiry', {
      method: 'POST',
      body: JSON.stringify(payload),
    });
  }
}

export class DelivezApiError extends Error {
  public statusCode: number;
  public details: any;

  constructor(statusCode: number, message: string, details?: any) {
    super(message);
    this.name = 'DelivezApiError';
    this.statusCode = statusCode;
    this.details = details;
  }
}
